### PR TITLE
feat(spanner): create instance using processing units/node count

### PR DESCRIPTION
**Pull request by jiren**
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-cloud-ruby/issues) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea.
- [x] Follow the instructions in [CONTRIBUTING](CONTRIBUTING.md). Most importantly, ensure the tests and linter pass by running `bundle exec rake ci` in the gem subdirectory.
- [x] Update code documentation if necessary.

closes: #11378

### CODE REVIEW

**Comment by jiren** [comment]
@hengfengli FYI, Generated gapic code is not added

File: google-cloud-spanner/lib/google/cloud/spanner/instance.rb
Comment (hengfengli) [inline google-cloud-spanner/lib/google/cloud/spanner/instance.rb:211]: lables -> labels

Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb]: done

File: google-cloud-spanner/lib/google/cloud/spanner/service.rb
Comment (hengfengli) [inline google-cloud-spanner/lib/google/cloud/spanner/service.rb:149]: This seems to be a breaking change. If the user doesn't provide a `field_mask` in the new version, will `update_instance` updates `display_name`, `node_count`, and `labels`?

Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/service.rb:149]: Consider to use `%w[display_name node_count labels]` as a default value of `field_mask`.

File: google-cloud-spanner/lib/google/cloud/spanner/project.rb
Comment (hengfengli) [inline google-cloud-spanner/lib/google/cloud/spanner/project.rb:249]: I'd like to avoid adding too many logic in the admin client, can we leave this check to the backend? So we don't need to add this check and its test as well.

File: google-cloud-spanner/lib/google/cloud/spanner/service.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/service.rb:149]: ok, I will make necessary changes.

File: google-cloud-spanner/lib/google/cloud/spanner/project.rb
Comment (hengfengli) [inline google-cloud-spanner/lib/google/cloud/spanner/project.rb:180]: Specifiy -> Specify

File: google-cloud-spanner/lib/google/cloud/spanner/project.rb
Comment (hengfengli) [inline google-cloud-spanner/lib/google/cloud/spanner/project.rb:182]: Same here.

File: google-cloud-spanner/lib/google/cloud/spanner/instance.rb
Comment (hengfengli) [inline google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: Can we provide a `field_mask` to `save` method? So we can avoid the change of line 246-255.

Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: This might not be a good idea. I was just trying to minimize the changes.

File: google-cloud-spanner/lib/google/cloud/spanner/instance.rb
Comment (hengfengli) [inline google-cloud-spanner/lib/google/cloud/spanner/instance.rb:251]: Remove this check as well. Let the backend to return the error and keep a thin client.

File: google-cloud-spanner/lib/google/cloud/spanner/instance.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: Yes we can provide optional argument for `field_mask`, but for usability wise it looks like it is not good. User need to first set the field and then need to pass field mask.

File: google-cloud-spanner/lib/google/cloud/spanner/service.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/service.rb:149]: Updated logic. Let me know your suggestions.

File: google-cloud-spanner/lib/google/cloud/spanner/project.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/project.rb]: done

File: google-cloud-spanner/lib/google/cloud/spanner/instance.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb]: done

File: google-cloud-spanner/lib/google/cloud/spanner/instance.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: As per new updated changes by default, it will updates `[display_name node_count labels]`.  If the user has updated fields then it will check for changes for respective fields and build field mask array,

Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: An issue with the current implementation is that users can't change the value twice. Let's say `processing_units` = 500 at the beginning, then we update it to `800`. After that, we change it again to `500` (this will fail because the value is equal to the original value `500`). You need to update `original_values` when the new values are saved.

File: google-cloud-spanner/lib/google/cloud/spanner/service.rb
Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/service.rb:149]: Looks good to me.

File: google-cloud-spanner/lib/google/cloud/spanner/instance.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: Updated original value after update request. One catch here is if a job fails then it will not revert back to original values.

Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: Is there any way that we don't store `original_values`? Another problem I notice is that if no thing changes, then `field_mask` will be empty. When we pass an empty array of `field_mask` to `update_instance`, which will use the default config `%w[display_name node_count labels]`.

Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: @hengfengli Is there any way to convert nodes to processing units? If there is a formula then calculate processing units and update only `processing_units`.

Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: > @hengfengli Is there any way to convert nodes to processing units? If there is a formula then we calculate processing units and update only `processing_units`.

No, I don't think it can work in this way.

Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: Actually, I am wrong. 1 node == 1000 processing_units. If we create 5000 processing unites, then it will be equal to 5 nodes.

Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: But I am a bit hesitate to hardcode this value into ruby client lib. I hope we don't do this.

Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: Can we remove `original_values`? and if `processing_units` is not empty, we use `processing_units`. If `processing_units` is empty, then we use `node_count`. What do you think?

Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: Processing units less than 1000 then api retuning 0 node count in response. If it is 1000 then returning node count is 1. 
If processing unit set to >1000 then node count is always present.

Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: Yeah, that's correct. But I don't want to write any thresholds in the code. So probably we just keep with the current `original_values` implementation. Instead of calling it `original_values`, may consider `current_values`.

Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: ok

Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/instance.rb:243]: renamed original values to `current_values`

**Comment by hengfengli** [comment]
@jiren Thanks a lot for fixing all the comments. It looks good to me now. I will approve it once it can be merged.

**Comment by hengfengli** [comment]
The proto change has been merged. We can start to merge this PR.

**Review by hengfengli — APPROVED** [review]

**Pull request merged**
